includes/Story_Renderer.php

```diff
@@ -0,0 +1,178 @@
+<?php
+/**
+ * Class Story_Renderer
+ *
+ * @package   Google\Web_Stories
+ * @copyright 2020 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://github.com/google/web-stories-wp
+ */
+
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Web_Stories;
+
+/**
+ * Class Story_Renderer
+ */
+class Story_Renderer {
+	/*
+	 * Regular expressions to fetch the individual structural tags.
+	 * These patterns were optimized to avoid extreme backtracking on large documents.
+	 */
+	const HTML_STRUCTURE_HTML_START_TAG = '/^(?<html_start>[^<]*(?:\s*<!--[^>]*>\s*)*<html(?:\s+[^>]*)?>)/i';
+
+	/**
+	 * Current post.
+	 *
+	 * @var \WP_Post Post object.
+	 */
+	protected $post;
+
+	/**
+	 * Story_Renderer constructor.
+	 *
+	 * @param \WP_Post $post Post object.
+	 */
+	public function __construct( $post ) {
+		$this->post = $post;
+	}
+
+	/**
+	 * Replaces the HTML start tag to make the language attributes dynamic.
+	 *
+	 * @param string $content Story markup.
+	 *
+	 * @return string Filtered content.
+	 */
+	protected function replace_html_start_tag( $content ) {
+		if ( preg_match( self::HTML_STRUCTURE_HTML_START_TAG, $content, $matches ) ) {
+			$replacement = sprintf( '<html amp %s>', get_language_attributes( 'html' ) );
+			$result      = preg_replace( self::HTML_STRUCTURE_HTML_START_TAG, $replacement, $content, 1 );
+
+			if ( null !== $result ) {
+				return $result;
+			}
+		}
+
+		return $content;
+	}
+
+	/**
+	 * Returns the full HTML <head> markup for a given story besides boilerplate.
+	 *
+	 * @return string Filtered content.
+	 */
+	protected function get_html_head_markup() {
+		ob_start();
+		?>
+		<style amp-custom>
+			.page-background-area, .page-safe-area {
```

Comment:
Since the CSS seems to match the CSS in JS, should these two perhaps come from a single file to avoid future duplicate editing?

Reply:
Hmm maybe. We could also remove this from PHP entirely.

The idea is that the JS-generated markup is fully working and AMP compatible. Also, the CSS might become dynamic in the future, so just reading from a static file wouldn't be enough.

Reply:
As mentioned above, will remove CSS from the PHP part for now, as it's static anyway.